### ISSUE TITLE
How to custom special point in webgl point layer

### ISSUE THREAD

**Issue by comictvn** [openlayers/openlayers#10381]
Hello anyone,

I have a request how to custom special point in WebGL Point Layer
In my understand WebGL just support LiteralStyle with some symbol type so I would like to draw special one like as RegularShape

Example for special icon like this
https://www.marinetraffic.com/en/ais/home/centerx:-28.3/centery:17.1/zoom:6

Please correct me if I'm wrong

Thanks!

**Comment by jahow** [comment]
You can specify an icon in a literal style object, I think this is the way to go in your case. See:
https://github.com/openlayers/openlayers/blob/fb73bbaa00227f0e45bda58d324a0127f259ae58/examples/webgl-points-layer.js#L15-L24

Unfortunately, **there is currently no support for rotation of icons in literal styles**. I'll mark this issue as PR accepted as it is an achievable goal IMO.

**Comment by comictvn** [comment]
@jahow Thank you so much
Is there any option to change icon color base on features's attributes

**Comment by jahow** [comment]
Sure, the `color` property will be applied as a tint on the icon. So a white square icon with a color of `red` will appear as a red square.

**Comment by comictvn** [comment]
`{
  "type":"Feature",
  "properties": {
    "shape":"sphere",
    "color":"#5aca5b"
  },
  "geometry": {
    "type":"Point",
    "coordinates":[0, 0]
  }
}`

@jahow so sorry for missing information
I mean each feature will have different color so how to change color base on property color

**Comment by jahow** [comment]
This should work then:
```
color: ['get', 'color'], 
```

The webgl styles are not part of the documented API yet, but you can find more info on expressions here:
https://github.com/openlayers/openlayers/blob/master/src/ol/style/expressions.js

**Comment by comictvn** [comment]
![Screen Shot 2019-12-06 at 4 23 25 PM](https://user-images.githubusercontent.com/6270931/70315291-d564fa80-184b-11ea-96a1-e8363506a5bc.png)

@jahow I got an error when was using `color: ['get', 'color']`

**Comment by jahow** [comment]
Indeed, this gives an error. Thanks for pointing that out!

What's missing is a way to interpret the features `color` attribute as a color rather than a string. Not sure what is the best way to go but this would definitely be a worthy improvement. I'll try to work on this in the weeks to come, in the meantime if you feel like contributing please be my guest :)

**Comment by comictvn** [comment]
Sure in next time :)
Thank you so much your kindly support

**Comment by comictvn** [comment]
**there is currently no support for rotation of icons in literal styles**
BTW this can be done by create 360 icon with different angel and use textureCoord

**Comment by delaneyj** [comment]
> Indeed, this gives an error. Thanks for pointing that out!
> 
> What's missing is a way to interpret the features `color` attribute as a color rather than a string. Not sure what is the best way to go but this would definitely be a worthy improvement. I'll try to work on this in the weeks to come, in the meantime if you feel like contributing please be my guest :)

I'm running into this exact issue with `['get', 'color']`.  It appears the shaders are compiled to `varying float v_color;`.  Obvious a float is a no go.  I wanted to create the vert/frag shader ala https://openlayers.org/workshop/en/webgl/points.html example but that API no longer works (fails looking for style).

**Comment by delaneyj** [comment]
`color: ['color', ['get', 'red'], ['get', 'green'], ['get', 'blue']],` appears to work but you have to separate the color into components

**Comment by jahow** [comment]
>  It appears the shaders are compiled to varying float v_color;

Exactly, strings are "converted" to numbers when sent to the GPU. It's useful for comparison but not for conversion to color. I'm not sure yet what the best solution to this would be.

You can still use the webgl renderer without literal styles though, as in the [heatmap layer](https://github.com/openlayers/openlayers/blob/master/src/ol/layer/Heatmap.js) for example. This will give you more flexibility. You could for example define 3 attributes (red, green and blue), each of them derived from the same feature property.

**Comment by baranyukVP** [comment]
Hi!
Can I customize WebGL points depends on map zoom?

**Comment by jahow** [comment]
Yes you can use the `["zoom"]` operator. See:

See the documentation here (should be available in the apidoc soon):
https://github.com/openlayers/openlayers/blob/master/src/ol/style/expressions.js
